google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolStressTest.java

```diff
@@ -0,0 +1,271 @@
+              }
+            });
+  }
+
+  private void setupSession(final Session session) {
+    ReadContext mockContext = mock(ReadContext.class);
+    ResultSet mockResult = mock(ResultSet.class);
+    when(session.singleUse(any(TimestampBound.class))).thenReturn(mockContext);
+    when(mockContext.executeQuery(any(Statement.class))).thenReturn(mockResult);
+    when(mockResult.next()).thenReturn(true);
+    doAnswer(
+            new Answer<Void>() {
+
+              @Override
+              public Void answer(InvocationOnMock invocation) throws Throwable {
+                synchronized (lock) {
+                  if (expiredSessions.contains(session.getName())) {
+                    throw SpannerExceptionFactory.newSpannerException(
+                        ErrorCode.NOT_FOUND, "Session not found");
+                  }
+                  if (sessions.remove(session.getName()) == null) {
+                    setFailed(closedSessions.get(session.getName()));
+                  }
+                  closedSessions.put(session.getName(), new Exception("Session closed at:"));
+                  if (sessions.size() < minSessionsWhenSessionClosed) {
+                    minSessionsWhenSessionClosed = sessions.size();
+                  }
+                }
+                return null;
+              }
+            })
+        .when(session)
+        .close();
+
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) throws Throwable {
+                if (random.nextInt(100) < 10) {
+                  expireSession(session);
+                  throw SpannerExceptionFactory.newSpannerException(
+                      ErrorCode.NOT_FOUND, "Session not found");
+                }
+                synchronized (lock) {
+                  if (sessions.put(session.getName(), true) != false) {
+                    setFailed();
+                  }
+                }
+                return null;
+              }
+            })
+        .when(session)
+        .prepareReadWriteTransaction();
+  }
+
+  private void expireSession(Session session) {
+    synchronized (lock) {
+      sessions.remove(session.getName());
+      expiredSessions.add(session.getName());
+    }
+  }
+
+  private void assertWritePrepared(Session session) {
+    synchronized (lock) {
+      if (!sessions.get(session.getName())) {
+        setFailed();
+      }
+    }
+  }
+
+  private void resetTransaction(Session session) {
+    synchronized (lock) {
+      sessions.put(session.getName(), false);
+    }
+  }
+
+  private void setFailed(Exception cause) {
+    e = new Exception(cause);
+  }
+
+  private void setFailed() {
+    e = new Exception();
+  }
+
+  private Exception getFailedError() {
+    synchronized (lock) {
+      return e;
+    }
+  }
+
+  @Test
+  public void stressTest() throws Exception {
+    int concurrentThreads = 10;
+    final int numOperationsPerThread = 1000;
```

Comment:
I think this makes sense to be in the test method as it is very specific to this test case.